### PR TITLE
Epoll ET AutoRead support

### PR DESCRIPTION
**Pull request by blucas**
Motivation:
Currently when using Epoll in Edge Triggered mode, autoread is not supported.

Modifications:
When autoRead is enabled and ```EpollMode``` is set to ```EpollMode.EDGE_TRIGGERED```, call ```epollInReady()``` directly to force a read from the socket

Result:
AutoRead functionality can now be supported in Edge Triggered mode.

### CODE REVIEW

**Comment by blucas** [comment]
@normanmaurer - As discussed, please take a look.

I consider this PR a **very rough draft**. Please let me know what I might have missed or any way I can simplify what I've done.

@Scottmitch - Don't know if this interests you, but maybe you'd like to take a look as well.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java [hunk 1]
```diff
@@ -824,7 +824,7 @@ void epollInReady() {
                     allocHandle.incMessagesRead(1);
                     pipeline.fireChannelRead(byteBuf);
                     byteBuf = null;
-                } while (allocHandle.continueReading());
+                } while (config.isAutoRead() && allocHandle.continueReading());
```
Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java]: can we put this in the `EpollRecvByteAllocatorHandle` implementations?

Comment (blucas) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java]: I'm glad you picked this one out. I do want to change this, but I'm not sure what I need to do... There are a few implementations of ```RecvByteBufAllocator``` and I'd like @normanmaurer to give me some guidance on how to include it correctly.

Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java]: and also add to the comments in these classes to indicate why this is necessary.

Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java]: auto read, and other channel options, and controlling characteristics of the read loop belong to `RecvByteBufAllocator` implementations. For `EPOLL` there are a few special conditions that must be enforced and so `EpollRecvByteAllocatorHandle` exists to wrap any existing `RecvByteBufAllocator.Handle` to apply the special epoll conditions on top of the typical behavior.  

Lets look at implementations of `EpollRecvByteAllocatorHandle`:
1. `EpollRecvByteAllocatorMessageHandle` - The autoRead channel option will be enforced by the delegate `super.continueReading()` ... and if EDGE triggered we must always read until there is no more data ... so this is effectively a form of forced "auto read" while in the read loop.
2. `EpollRecvByteAllocatorStreamingHandle` - Again the autoRead channel option will be enforced by the delegate `super.continueReading()` but for EPOLL we continue reading as long as the last read operation actually read as much data as we requested ... if this is the case there more be more data to read. This is a slightly less strong form of "forcing auto read" but it seems to be doing the same thing.

So in summary I'm not sure why this additional call to `config.isAutoRead()` is necessary ... can you explain the scenario you are in (or create a unit tests) where this is necessary?

Comment (blucas) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java]: I agree that the logic for determining that the read loop should exit should live in the ```RecvByteBufAllocator.DelegatingHandle```.  This is just a very rough draft to prove what might allow us to provide *normal behaviour* for autoRead in EDGE mode.

The ```RecvByteBufAllocator.DelegatingHandle``` Epoll implementations will completely disregard the value of ```config.isAutoRead()``` when in EDGE mode (because of the logic [here](https://github.com/netty/netty/blob/4.1/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorMessageHandle.java#L37) and [here](https://github.com/netty/netty/blob/4.1/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorStreamingHandle.java#L39)). So to get around that, I added the ```config.isAutoRead()``` check to the read loop.  This way if an application's ```ChannelHandler``` sets ```autoRead``` to ```false``` while executing ```channelRead``` the transport (in this specific case, EpollMode.ET) will honour that and break out of the loop.

Again I know this is not the right place to put the check, the check should be in the ```RecvByteBufAllocator``` implementations.  I would like to get @normanmaurer to give me some tips on the best way to accomplish that.

Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java]: > The RecvByteBufAllocator.DelegatingHandle Epoll implementations will completely disregard the value of config.isAutoRead() when in EDGE mode

@blucas - Auto read is intentionally ignored because of the way EPOLL ET works. If ET is enabled and we don't read all available data, then we may never be notified again there is data pending (see [Level-triggered and edge-triggered](http://man7.org/linux/man-pages/man7/epoll.7.html)).

**Comment by blucas** [comment]
@normanmaurer - gentle ping :)

**Comment by normanmaurer** [comment]
Sorry super busy will try to find some time tomorrow

**Comment by Scottmitch** [comment]
@blucas - I think the `moreData` idea we discussed offline needs to be present, and we need to check this in the `read()` call path. Otherwise if a user switches from `autoread=true` to `autoread=false` in a read loop, and then at some later time calls `read()` EPOLL will never notify us that there is more data. So somewhere in the `AbstractEpollChannel.doBeginRead()` path.

**Comment by Scottmitch** [comment]
@blucas - I think I have a handle on this ... let me take this over. I'll ping you for a review when I have a PR.

**Comment by Scottmitch** [comment]
Closed in favor of https://github.com/netty/netty/pull/4919
